### PR TITLE
librbd: don't report HOLE_UPDATED when diffing against a hole

### PR DESCRIPTION
**Pull request by idryomov**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by idryomov** [comment]
No related failures:

https://pulpito.ceph.com/dis-2023-12-12_01:08:53-rbd-wip-dis-testing-2-distro-default-smithi/
https://pulpito.ceph.com/dis-2023-12-12_12:31:55-rbd-wip-dis-testing-2-distro-default-smithi/

**Comment by idryomov** [comment]
Reordered `TestMockObjectMapDiffRequest` tests, no functional change.

**Comment by idryomov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
@trociny This should be ready, I would appreciate you taking a look if you have some time.

File: src/librbd/object_map/DiffRequest.cc [hunk 1]
```diff
@@ -195,16 +195,34 @@ void DiffRequest<I>::handle_load_object_map(int r) {
...
+        *diff_it = DIFF_STATE_DATA_UPDATED;
+      }
+      break;
+    case DIFF_STATE_DATA_UPDATED:
+      if (object_map_state == OBJECT_NONEXISTENT) {
+        *diff_it = DIFF_STATE_HOLE_UPDATED;
+      }
+      break;
+    default:
+      ceph_abort();
```
Comment (trociny) [inline src/librbd/object_map/DiffRequest.cc:225]: I like this refactoring. It looks much clear now.

**Review by trociny — APPROVED** [review]
LGTM.

Initially I thought that I didn't like very much `diff_iterate_range` name as a param for `object_map::DiffRequest`. For me it was not clear what it meant until I looked at the code, and I would prefer either "reverse" param `deep_copy` to indicate that it was used for specific deep copy case, or much better if the param name would tell what actually it would do instead of what caller made the request, i.e. something like `{skip|keep}_unnecessary_holes` (I can't think of a good name though :-) But then seeing the "diff iterate"/"deep copy" terms used in tests I got used to it. So now I am fine with this name too.

**Comment by idryomov** [comment]
> Initially I thought that I didn't like very much `diff_iterate_range` name as a param for `object_map::DiffRequest`.

The "range" in the name refers to

> - diff is performed on the entire image no matter how small the byte range is, unused results just get discarded before returning

bullet in https://tracker.ceph.com/issues/63341.  I envision `diff_iterate_range` becoming `std::optional<std::pair<uint64_t, uint64_t>` for start and end object numbers.  `std::nullopt` would signify deep-copy, which is not ranged by definition.

Thanks for the review!

**Pull request merged**